src/Plugin/Field/FieldType/OgMembershipItemList.php

```diff
@@ -0,0 +1,191 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\og\Plugin\Field\FieldType\OgMembershipItemList.
+ */
+
+namespace Drupal\og\Plugin\Field\FieldType;
+
+use Drupal\Core\Field\EntityReferenceFieldItemList;
+use Drupal\og\Controller\OG;
+use Drupal\og\Entity\OgMembership;
+
+/**
+ * Defines a item list class for OG membership fields.
+ */
+class OgMembershipItemList extends EntityReferenceFieldItemList {
+
+  /**
+   * Whether or not this list of field items has fetched membership groups.
+   *
+   * @var bool
+   */
+  protected $fetched = FALSE;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function count() {
+    // Overrides ListItem::count, implementing \Countable::count
+    // Fetch the list of memberships for this field.
+    if (!$this->fetched) {
+      $this->populateGroupsFromMembershipEntities();
+      $this->fetched = TRUE;
+    }
+
+    return parent::count();
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @todo Need to account for new items added and not being overridden etc..
+   */
+  public function get($index) {
+    if (!is_numeric($index)) {
+      throw new \InvalidArgumentException('Unable to get a value with a non-numeric delta in a list.');
+    }
+
+    // Fetch the list of memberships for this field.
+    if (!$this->fetched) {
+      $this->populateGroupsFromMembershipEntities();
+      $this->fetched = TRUE;
+    }
+
+    // Automatically create the first item for computed fields.
+    if ($index == 0 && !isset($this->list[0])) {
+      $this->list[0] = $this->createItem(0);
+    }
+
+    return isset($this->list[$index]) ? $this->list[$index] : NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function postSave($update) {
+    parent::postSave($update);
+
+    $group_ids = [];
+    foreach ($this->list as $item) {
+      $group_ids[] = $item->getValue()['target_id'];
+    }
+
+    // todo: move to API function.
+    $membership_ids = \Drupal::entityQuery('og_membership')
+      ->condition('etid', $this->getEntity()->id())
+      ->condition('entity_type', $this->getEntity()->getEntityTypeId())
+      ->condition('group_type', $this->getFieldDefinition()->getTargetEntityTypeId())
+      ->condition('field_name', $this->getFieldDefinition()->getName())
+      ->execute();
+
+    /** @var \Drupal\og\Entity\OgMembership[] $memberships */
+    $memberships = OgMembership::loadMultiple($membership_ids);
+
+    $target_group_ids = array_map(function($membership) {
+      return $membership->getGid();
+    }, $memberships);
+
+    $deprecated_membership_ids = array_diff($target_group_ids, $group_ids);
+    $new_membership_group_ids = array_diff($group_ids, $target_group_ids);
+
+// @todo What was this doing before, confused.
+//      if (!$target_group_ids) {
```

Comment:
ping @RoySegall I was not sure what was going on here, as target_ids could also be empty if there are no memberships yet? This would already be handled by the diffing.

I guess I need you to explain how you have these 'orphans'. sorry :)